kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/seek/IndexManager.scala

```diff
@@ -0,0 +1,184 @@
+     * @param maybeIndex option of the index filename
+     * @return either an error, or a TopicPartitionOffset
+     */
+    def indexToOffset(maybeIndex: Option[String]): Either[Throwable, Option[TopicPartitionOffset]] = {
+      maybeIndex match {
+        case Some(index) =>
+          for {
+            offset <- IndexFilenames.offsetFromIndex(index)
+          } yield Some(topicPartition.withOffset(offset))
+        case None => Option.empty[TopicPartitionOffset].asRight
+      }
+    }
+
+    storageInterface.list(bucketAndPrefix.withPath(
+      IndexFilenames.indexForTopicPartition(sinkName, topicPartition.topic.value, topicPartition.partition)
+    ))
+      .leftMap(e => new NonFatalS3SinkError("Couldn't retrieve listing", e.exception))
+      .flatMap(
+        indexes =>
+          if (indexes.size > maxIndexes) {
+            logAndReturnMaxExceededError(topicPartition, indexes)
+          } else if (indexes.isEmpty && fallbackSeeker.nonEmpty) {
+            fallbackOffsetSeek(fallbackSeeker.get, topicPartition, fileNamingStrategy, bucketAndPrefix)
+          } else {
+            {
+              for {
+                validIndex <- scanIndexes(bucketAndPrefix, indexes)
+                indexesToDelete = indexes.filterNot(validIndex.contains)
+                _ <- storageInterface.deleteFiles(bucketAndPrefix.bucket, indexesToDelete)
+                offset <- indexToOffset(validIndex)
+              } yield {
+                logger.info("[{}] Seeked offset {} for TP {}", sinkName, offset, topicPartition)
+                offset
+              }
+            }.leftMap {
+              case err: FileLoadError =>
+                val logLine = s"File load error while seeking: ${err.message()}"
+                logger.error(s"[{}] {}",sinkName, logLine, err.exception)
+                new NonFatalS3SinkError(logLine, err.exception)
+              case err: FileDeleteError =>
+                val logLine = s"File delete error while seeking: ${err.message()}"
+                logger.error(s"[{}] {}",sinkName, logLine, err.exception)
+                new NonFatalS3SinkError(logLine, err.exception)
+              case err: Throwable =>
+                val logLine = s"Error while seeking: ${err.getMessage}"
+                logger.error(s"[{}] {}",sinkName, logLine, err)
+                new NonFatalS3SinkError(logLine, err)
+            }
+          }
+      )
+  }
+
+  /**
+   * Given a bucket and a list of files, attempts to load them to establish the most recent valid index
+   *
+   * @param bucketAndPrefix Remote S3 Root Location of the sink files
+   * @param indexFiles      List of index files
+   * @return either a FileLoadError or an optional string containing the valid index file of the offset
+   */
+  def scanIndexes(bucketAndPrefix: RemoteS3RootLocation, indexFiles: List[String]): Either[FileLoadError, Option[String]] = {
+    var latest: Either[FileLoadError, Option[String]] = None.asRight
+
+    indexFiles
```

Comment:
would it be better to translate to long the index file name and then sort on that?

Reply:
I think this is marginally more efficient.

With AWS S3 there is a guarantee that [list results are always returned in UTF-8 binary order](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html).

As long as we are sure the index files are correctly padded we can rely on the order they are returned from S3.

Which means fewer in-memory sort operations.

For the number of results we're talking about it wouldn't make much difference but I'm thinking it seems to be a good practice.